cmake/external/grpc.cmake

```diff
@@ -12,119 +12,123 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-include(ExternalProject)
+if(TARGET grpc)
+  return()
+endif()
+
 include(ExternalProjectFlags)
+include(external/c-ares)
+include(external/protobuf)
+include(external/zlib)
 
 if(GRPC_ROOT)
   # If the user has supplied a GRPC_ROOT then just use it. Add an empty custom
   # target so that the superbuild dependencies still work.
   add_custom_target(grpc)
-
-else()
-  set(
-    GIT_SUBMODULES
-    third_party/boringssl
-  )
-
-  set(
-    CMAKE_ARGS
-    -DCMAKE_BUILD_TYPE:STRING=${CMAKE_BUILD_TYPE}
-    -DBUILD_SHARED_LIBS:BOOL=OFF
-    -DgRPC_BUILD_TESTS:BOOL=OFF
-
-    # TODO(rsgowman): We're currently building nanopb twice; once via grpc, and
-    # once via nanopb. The version from grpc is the one that actually ends up
-    # being used. We need to fix this such that either:
-    #   a) we instruct grpc to use our nanopb
-    #   b) we rely on grpc's nanopb instead of using our own.
-    # For now, we'll pass in the necessary nanopb cflags into grpc. (We require
-    # 16 bit fields. Without explicitly requesting this, nanopb uses 8 bit
-    # fields.)
-    -DCMAKE_C_FLAGS=-DPB_FIELD_16BIT
-    -DCMAKE_CXX_FLAGS=-DPB_FIELD_16BIT
-  )
-
-
-  ## c-ares
-  if(NOT c-ares_DIR)
-    set(c-ares_DIR ${FIREBASE_INSTALL_DIR}/lib/cmake/c-ares)
-  endif()
-
-  list(
-    APPEND CMAKE_ARGS
-    -DgRPC_CARES_PROVIDER:STRING=package
-    -Dc-ares_DIR:PATH=${c-ares_DIR}
-  )
-
-
-  ## protobuf
-
-  # Unlike other dependencies of gRPC, we control the protobuf version because we
-  # have checked-in protoc outputs that must match the runtime.
-
-  # The location where protobuf-config.cmake will be installed varies by platform
-  if(NOT Protobuf_DIR)
-    if(WIN32)
-      set(Protobuf_DIR "${FIREBASE_INSTALL_DIR}/cmake")
-    else()
-      set(Protobuf_DIR "${FIREBASE_INSTALL_DIR}/lib/cmake/protobuf")
-    endif()
+  return()
+endif()
```

Comment:
cmake files often repeat the if condition in the endif. (The old code did that here, though the block was much larger too.) For consistency's sake, you *might* want to add that back in.

Personally, I think it's largely needless boilerplate. (But I feel the same way about the `}  // namespace boilerplate` comments. shrug.)

(highly) optional. Again, if you do choose to do this, do so throughout.

Reply:
CMake used to require the condition to match in the trailing `endif` but no longer does. For short blocks I don't think the noise is worth it. I've left a few instances in here where the blocks are large and contain additional branching (usually the final block in the FindFoo.cmake files).